shell/platform/windows/win32_window.cc

```diff
@@ -190,35 +194,58 @@ Win32Window::MessageHandler(HWND hwnd,
         // DefWindowProc will send WM_CHAR for this WM_UNICHAR.
         break;
       }
+      case WM_DEADCHAR:
+      case WM_SYSDEADCHAR:
       case WM_CHAR:
       case WM_SYSCHAR: {
-        if (wparam == VK_BACK)
-          break;
         char32_t code_point = static_cast<char32_t>(wparam);
         static char32_t lead_surrogate = 0;
-        // If code_point is LeadSurrogate, save and return.
+        // If code_point is LeadSurrogate, save to combine to potentially form
+        // a complex Unicode character.
         if ((code_point & 0xFFFFFC00) == 0xD800) {
           lead_surrogate = code_point;
-          return TRUE;
         }
         // Merge TrailSurrogate and LeadSurrogate.
         if (lead_surrogate != 0 && (code_point & 0xFFFFFC00) == 0xDC00) {
           code_point = 0x10000 + ((lead_surrogate & 0x000003FF) << 10) +
                        (code_point & 0x3FF);
         }
         lead_surrogate = 0;
-        window->OnChar(code_point);
+        if (wparam != VK_BACK) {
+          window->OnChar(code_point);
+        }
+
+        // All key presses that generate a character should be sent from
+        // WM_CHAR. In order to send the full key press information, the keycode
+        // is persisted in a static variable keycode_for_char_message obtained
+        // from WM_KEYDOWN.
+        const unsigned int scancode = (lparam >> 16) & 0xff;
+        window->OnKey(keycode_for_char_message, scancode, WM_KEYDOWN, code_point);
```

Comment:
I'm confused about the complex character flow. First we send the initial code point, then later we send the combined code point? And sending the first part twice, effectively, doesn't break anything?

If that's correct behavior, can you maybe add a comment pointing to the relevant framework code explaining/showing why?

Reply:
Just added a check to avoid sending the dead key code point, and some comments explaining the flow. Is this better?

Reply:
The dead key check seems different. Or is a dead key always a leading surrogate?

Reply:
Capturing from offline discussion: dead keys and surrogate pairs are orthogonal. The decision was that we should add a check to only send OnKey once even if we get multiple WM_CHAR messages, which should address my concern here.